### PR TITLE
Fixed NeptuneLogger when using DDP

### LINKED ISSUE #10604: When using neptune logger and multi-gpu DDP, duplicate runs are logged, one for each gpu.
## 🐛 Bug

With using the neptune logger and trainer with n_gpu > 1, accelerator='ddp', duplicate experiments are created and logged, one for each individual gpu

### To Reproduce

```py
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer

from pytorch_lightning.loggers.neptune import NeptuneLogger

# Create NeptuneLogger
neptune_logger = NeptuneLogger(
    project="m-bain/LONGVID")


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        gpus=4,
        strategy='ddp',
        logger=neptune_logger
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()
```

Output:

```
https://app.neptune.ai/m-bain/LONGVID/e/LON-52
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
https://app.neptune.ai/m-bain/LONGVID/e/LON-53
initializing distributed: GLOBAL_RANK: 1, MEMBER: 2/4
https://app.neptune.ai/m-bain/LONGVID/e/LON-54
initializing distributed: GLOBAL_RANK: 2, MEMBER: 3/4
initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/4
https://app.neptune.ai/m-bain/LONGVID/e/LON-55
initializing distributed: GLOBAL_RANK: 3, MEMBER: 4/4
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All distributed processes registered. Starting with 4 processes
----------------------------------------------------------------------------------------------------
```

### Expected behavior

A single run logged (happens when gpu=1 and/or accelerator is turned off), not 4 of them...

### Environment

- PyTorch Lightning Version (e.g., 1.3.0): 1.5.1
- PyTorch Version (e.g., 1.8): 1.10.0
- Python version: 3.8.12
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: cuda 11.3
- GPU models and configuration: 4 x NVIDIA GeForce GTX TITAN X
- How you installed PyTorch (`conda`, `pip`, source): pip

cc @tchaton

### PR DESCRIPTION
**Pull request by Raalsky**
## What does this PR do?

Makes Neptune Experiment (Run in the scope of Neptune) initialization to be called only on rank == 0 in distributed environment. Should not introduce any breaking changes and was tested with previously pickled `NeptuneLogger`.

Fixes #10604

And could resolve a few similar issues in external repositories:
- https://github.com/neptune-ai/neptune-client/issues/733
- https://github.com/neptune-ai/neptune-pytorch-lightning/issues/12
- https://github.com/neptune-ai/neptune-pytorch-lightning/issues/11

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃
_Definitely ❤️_

### CODE REVIEW

**Review by shnela — APPROVED** [review]

File: pytorch_lightning/loggers/neptune.py [hunk 1]
```diff
@@ -412,8 +421,24 @@ def training_step(self, batch, batch_idx):
...
+                self._run_instance[_INTEGRATION_VERSION_KEY] = __version__
+
+            return self._run_instance
+        except NeptuneLegacyProjectException as e:
+            raise TypeError(
+                f"""Project {self._project_name} has not been migrated to the new structure.
+                You can still integrate it with the Neptune logger using legacy Python API
+                available as part of neptune-contrib package:
+                  - https://docs-legacy.neptune.ai/integrations/pytorch_lightning.html\n
+                """
```
Comment (carmocca) [inline pytorch_lightning/loggers/neptune.py:440]: Don't use triple quotes as they include the indentation as leading spaces.

```suggestion
                f"Project {self._project_name} has not been migrated to the new structure."
                " You can still integrate it with the Neptune logger using legacy Python API"
                " available as part of neptune-contrib package:"
                " https://docs-legacy.neptune.ai/integrations/pytorch_lightning.html\n
                "
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loggers/neptune.py [hunk 2]
```diff
@@ -273,44 +273,53 @@ def __init__(
...
-        # make sure that we've log integration version for both newly created and outside `Run` instances
-        run_instance[_INTEGRATION_VERSION_KEY] = __version__
-
-        # keep api_key and project, they will be required when resuming Run for pickled logger
-        self._api_key = api_key
-        self._project_name = run_instance._project_name  # skipcq: PYL-W0212
+    @property
+    def _neptune_init_args(self):
+        args = {}
+        # Backward compatibility in case of previous version retrieval
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/neptune.py:300]: can you put the exact version here?

Comment (rohitgr7) [reply pytorch_lightning/loggers/neptune.py:300]: in case it is 0.9, you can use `_NEPTUNE_GREATER_EQUAL_0_9`

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/loggers/neptune.py (hunk 2 above)
Comment (Raalsky) [reply pytorch_lightning/loggers/neptune.py:300]: It's not due to Neptune incompatibility. It's only about pickle-unpickle from previous pytorch-lightning version :wink:

Comment (Raalsky) [reply pytorch_lightning/loggers/neptune.py:300]: Actually we only need a "run id" to recover an already pickled Neptune run + maybe project info + api_token and all that data are stored in previously pickled (for instance pickled with current PTL master) states. Now we need to store a little bit more data that is used on the initialization step of Neptune Run but it's still possible to reuse old pickled loggers and reuse old Neptune runs.

It's only for the purpose of beeing consider as bugfix instead of behavioral change 😎

Comment (rohitgr7) [reply pytorch_lightning/loggers/neptune.py:300]: oh! I see. thanks for clarification :)

**Review by awaelchli — APPROVED** [review]
Great! Thanks for the quick fix!

One could consider adding Neptune to the parametrization of 
`tests/loggers/test_all.py::test_logger_created_on_rank_zero_only`

**Comment by Raalsky** [comment]
> Great! Thanks for the quick fix!
> 
> One could consider adding Neptune to the parametrization of `tests/loggers/test_all.py::test_logger_created_on_rank_zero_only`

Added :wink:. Sorry, am I broke something (don't know about an active auto merge before committing)? :D

File: pytorch_lightning/loggers/neptune.py [hunk 3]
```diff
@@ -273,44 +273,53 @@ def __init__(
         super().__init__()
         self._log_model_checkpoints = log_model_checkpoints
         self._prefix = prefix
+        self._run_name = name
```
Comment (awaelchli) [inline pytorch_lightning/loggers/neptune.py:276]: Can't push to the branch for some reason. Can you add this code at line 269 please?
```py
        if neptune is None:
            raise ModuleNotFoundError(
                "You want to use the `Neptune` logger which is not installed yet, install it with"
                " `pip install neptune-client`."
            )
```

This should fix the remaining issue and I checked locally the tests are running. Thanks

Comment (Raalsky) [reply pytorch_lightning/loggers/neptune.py:281]: Thank you so much :wink:. I've updated the logger initializer as you wish and I confirm that it passes locally :wink:

File: CHANGELOG.md [hunk 4]
```diff
@@ -308,7 +308,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed support for `CombinedLoader` while checking for warning raised with eval dataloaders ([#10994](https://github.com/PyTorchLightning/pytorch-lightning/pull/10994))
 - The TQDM progress bar now correctly shows the `on_epoch` logged values on train epoch end ([#11069](https://github.com/PyTorchLightning/pytorch-lightning/pull/11069))
 - Fixed bug where the TQDM updated the training progress bar during `trainer.validate` ([#11069](https://github.com/PyTorchLightning/pytorch-lightning/pull/11069))
-
+- Fixed `NeptuneLogger` when using DDP ([#11030](https://github.com/PyTorchLightning/pytorch-lightning/pull/11030))
```
Comment (rohitgr7) [inline CHANGELOG.md:311]: you might need to move the changelog to the unreleased section.

Comment (Raalsky) [reply CHANGELOG.md]: Sorry. I was too optimistic during conflicts resolving. Is it ok now?

File: CHANGELOG.md [hunk 5]
```diff
@@ -309,7 +311,6 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - The TQDM progress bar now correctly shows the `on_epoch` logged values on train epoch end ([#11069](https://github.com/PyTorchLightning/pytorch-lightning/pull/11069))
 - Fixed bug where the TQDM updated the training progress bar during `trainer.validate` ([#11069](https://github.com/PyTorchLightning/pytorch-lightning/pull/11069))
```
Comment (rohitgr7) [inline CHANGELOG.md:313]:
```suggestion


```

Comment (rohitgr7) [reply CHANGELOG.md:313]: weird.. for some reason.. can't push commits here 🤔

Comment (Raalsky) [reply CHANGELOG.md]: I think the main reason is that the source repository is in `neptune-ai` organization

**Review by Borda — APPROVED** [review]

**Comment by Raalsky** [comment]
It looks like a final state now :wink:. @rohitgr7 @awaelchli :wink: Could you re-enable auto-merge?
